### PR TITLE
Voting WRITE side and api logic

### PR DESCRIPTION
**Pull request by paweljakubas**
- [x]  write side of certs (from wallet to ledger)
- [x] logic - is conway era
- [x] logic for withdrawals in conway
- [x] vote to abstain logic    

### Issue Number
adp-3212

### CODE REVIEW

File: lib/wallet/src/Cardano/Wallet.hs [hunk 1]
```diff
@@ -2623,6 +2630,76 @@ submitExternalTx tr nw tl sealedTx = do
...
+    -> ExceptT ErrConstructTx IO ()
+votingEraValidation nw = do
+    era <- liftIO $ currentNodeEra nw
+    case era of
+        Cardano.AnyCardanoEra Cardano.ConwayEra -> pure ()
+        _ -> throwE ErrConstructTxVotingInWrongEra
+
+checkingIfVoted
+    :: DBLayer IO s
+    -> NetworkLayer IO block
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet.hs]:
```suggestion
    -> RecentEra era
```

You have an `era :: RecentEra era` very close-by at the call-sites! This way there's only one source of truth and no potential race conditions from reading the era twice.

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs [hunk 2]
```diff
@@ -486,6 +486,18 @@ instance IsServerError ErrConstructTx where
...
+            [ "I cannot construct a transaction that includes voting before "
+            , "the Conway era."
+            ]
+        ErrConstructTxWithdrawalWithoutVoting ->
+            apiError err403 WithdrawalNotPossibleWithoutVote $ mconcat
+            [ "I cannot construct a transaction that contains withdrawals "
+            , "if voting was not cast in the Conway era onwards. Please re-delegate "
+            , "(in that case abstain vote will be cast) or make a vote transaction "
+            , "before the withdrawal transaction."
+            ]
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs]:
```suggestion
            [ "I cannot construct a transaction that contains withdrawals "
            , "if the wallet is not voting in the Conway era onwards. "
            , "Please delegate again (in that case, the wallet will automatically vote to abstain), "
            , "or make a vote transaction before the withdrawal transaction."
            ]
```

File: lib/wallet/src/Cardano/Wallet.hs [hunk 3]
```diff
@@ -2623,6 +2630,76 @@ submitExternalTx tr nw tl sealedTx = do
...
     pure txid
 
+votingEraValidation
+    :: NetworkLayer IO block
+    -> ExceptT ErrConstructTx IO ()
+votingEraValidation nw = do
+    era <- liftIO $ currentNodeEra nw
+    case era of
+        Cardano.AnyCardanoEra Cardano.ConwayEra -> pure ()
+        _ -> throwE ErrConstructTxVotingInWrongEra
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet.hs]:
```suggestion
votingEnabledInEra
    :: RecentEra era
    -> Either ErrConstructTx ()
votingEnabledInEra = \case
    RecentEraConway -> Right ()
    _ -> Left ErrConstructTxVotingInWrongEra
```

This logic is best expressed as a pure function.

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 4]
```diff
@@ -3275,13 +3292,19 @@ constructSharedTransaction
     delegationRequest <-
         liftHandler $ traverse parseDelegationRequest $ body ^. #delegations
 
+    when (isJust (body ^. #vote)) $
+        liftHandler $ W.votingEraValidation nl
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
        liftHandler $ except $ W.votingEnabledInEra era
```

In order to have `era` in scope, you need to move the check inside the argument to `withWorkerCtx`, after the call to the `readNodeTipStateForTxWrite` function.

Moving the validity checks on the POST request body inside the argument of `withWorkerCtx` is a good idea anyway, since we first want to check that the wallet with `wid` exists, and only then start inspecting the transaction body.

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 5]
```diff
@@ -2722,6 +2722,9 @@ constructTransaction api argGenChange knownPools poolStatus apiWalletId body = d
                 } -> liftHandler $ throwE ErrConstructTxWrongPayload
             _ -> pure ()
 
+    when (isJust (body ^. #vote)) $
+        liftHandler $ W.votingEraValidation nl
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
        liftHandler $ except $ W.votingEnabledInEra era
```

See the other comment.

File: lib/wallet/src/Cardano/Wallet.hs [hunk 6]
```diff
@@ -2623,6 +2630,76 @@ submitExternalTx tr nw tl sealedTx = do
...
+                throwE ErrConstructTxWithdrawalWithoutVoting
+        _ -> pure ()
+
+alreadyVoted
+    :: Functor stm
+    => DBVar stm (DeltaWalletState s)
+    -> stm Bool
+alreadyVoted walletState =
+    Dlgs.isVoting . view #delegations
+        <$> readDBVar walletState
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet.hs]:
```suggestion
assertIsVoting
    :: DBLayer IO s
    -> RecentEra era
    -> ExceptT ErrConstructTx IO ()
assertIsVoting DBLayer{..} = \case
    RecentEraConway -> do
        voting <- liftIO . atomically . fmap isVoting $ readDBVar walletState
        if voting then
            pure ()
        else
            throwE ErrConstructTxWithdrawalWithoutVoting
    _ -> pure ()

isVoting :: WalletState -> Bool
isVoting = Dlgs.isVoting . view #delegations
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 7]
```diff
@@ -3275,13 +3292,19 @@ constructSharedTransaction
...
+
     withWorkerCtx @_ @_ @Handler api wid liftE liftE $ \wrk -> do
         let db = wrk ^. dbLayer
             netLayer = wrk ^. networkLayer
 
             trWorker :: Tracer IO W.WalletLog
             trWorker = MsgWallet >$< wrk ^. logger
 
+        when (isJust (body ^. #withdrawal)) $
+            liftHandler $ W.checkingIfVoted db netLayer
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
            liftHandler $ W.assertIsVoting db era
```

For getting `era` in scope, see one of the other comments.

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 8]
```diff
@@ -2750,6 +2753,9 @@ constructTransaction api argGenChange knownPools poolStatus apiWalletId body = d
...
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
            liftHandler $ W.assertIsVoting db era
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 9]
```diff
@@ -2774,11 +2781,20 @@ constructTransaction api argGenChange knownPools poolStatus apiWalletId body = d
                     db currentEpochSlotting knownPools
                     poolStatus withdrawal
 
+        optionalVoteAction <- case (body ^. #vote) of
+            Just (ApiT action) ->
+                liftIO $ Just <$> WD.voteAction trWorker db action
+            Nothing ->
+                liftIO $ W.handleVotingWhenMissingInConway nl db
+                              (isJust (body ^. #delegations))
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
            Nothing -> pure Nothing
```

⚠️ 

I have come to the conclusion that the "Vote to abstain" feature is best implemented as part of the `joinStakePool` endpoint. The reason is that

* The intent of `constructTransaction` is to construct a transaction from the given JSON data. It performs a bit of validation, but the JSON input is essentially a one-to-one representation of a Cardano transaction — no "magic".
* The intent of `joinStakePool` function is to perform the necessary steps to join a stake pool — this includes "magic" such as automatically voting to a `DRep`.

----

In general, the `Cardano.Wallet.Api.Http.Shelley.Server` module is not a good place for business logic such as `handleVotingWhenMissingInConway`. The main concern of this module is to parse and partially validate JSON data. We need to separate concerns properly in order to keep our code manageable. I'm afraid, but this module has grown way past manageable  — and this slows down development of new features and dependency updates.

File: lib/wallet/src/Cardano/Wallet.hs [hunk 10]
```diff
@@ -2623,6 +2630,76 @@ submitExternalTx tr nw tl sealedTx = do
...
+
+handleVotingWhenMissingInConway
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet.hs:2688]: As described in `Cardano.Wallet.Api.Http.Shelley.Server`, I do not wish `constructTransaction` to perform automatic voting. This means that this function `handleVotingWhenMissingInConway` should be removed / absorbed into the `joinStakePool` logic.

File: lib/wallet/src/Cardano/Wallet/Transaction/Delegation.hs [hunk 11]
```diff
@@ -53,19 +66,24 @@ certificateFromDelegationAction
         -- ^ Era in which we create the certificate
     -> Either XPub (Script KeyHash)
         -- ^ Our staking credential
-    -> DelegationAction
+    -> Maybe DelegationAction
         -- ^ Delegation action that we plan to take
+    -> Maybe VotingAction
+        -- ^ Optional vote action in Conway era onwards
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Transaction/Delegation.hs]:
```suggestion
    -> DelegationAction
        -- ^ Delegation action that we plan to take
```

Hm, this function is becoming a mess, I am afraid. 😵‍💫 @paweljakubas, could you:

* remove the logic for `Maybe VotingAction` from this function again
* and instead add a new module `Cardano.Wallet.Transaction.Voting` with an analogous function `certificateFromVotingAction`. Then, you can use this function to create certificates that are specific to voting and join them with the certificates for delegation — see other comment.

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 12]
```diff
@@ -711,11 +713,14 @@ mkUnsignedTransaction networkId stakeCred ctx selection = do
                         inpsScripts
                     Nothing refScriptM
         Just action -> do
+            let deposit = view #txDeposit ctx
             let certs = case stakeCred of
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs]:
```suggestion
            let delegationCerts = case stakeCred of …
```

```hs
            let deduplicateRegister (JoinRegisteringKey k) (VotingRegisteringKey v) = Voting v
                deduplicateRegister _ v = v
            let votingCerts = case deduplicateRegister <$> (#txVotingAction ctx) of … ⟨ certificateFromVotingAction ⟩
            let certs = delegationCerts ++ votingCerts 
```

Here, the idea is to separate two concerns: Adding delegation certificates and adding voting certificates. It's ok to use two certificates in a single transaction; the code becomes too much of a mess otherwise.

File: specifications/Cardano/Wallet/delegation.md
Comment (HeinrichApfelmus) [inline specifications/Cardano/Wallet/delegation.md:1]: Could you split this file off into a separate pull request?

Easier to review and change that way.

**Review by HeinrichApfelmus — COMMENTED** [review]
Thank you for the pull request!

First round of comments.

The most important request is that I do not wish the "delegate to abstain" logic be part of the `constructTransaction` endpoint — I want it to be part of the `joinStakePool` endpoint.

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs]: done

File: lib/wallet/src/Cardano/Wallet.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet.hs]: fair argument - done

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 4 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: used combinator from `Data.Either.Combinators` : 
```
whenLeft (W.votingEnabledInEra era) (liftHandler .throwE)
```

File: lib/wallet/src/Cardano/Wallet.hs (hunk 6 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet.hs]: done

File: lib/wallet/src/Cardano/Wallet.hs (hunk 1 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet.hs]: done

File: lib/wallet/src/Cardano/Wallet.hs (hunk 10 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet.hs:2688]: handleVotingWhenMissingInConway will incur `Abstain` only when 
1.we are in Conway era
2. we have not already voted
3. we are about to delegate in the same tx
I really thought this was "delegate to abstain" feature that was requested

When it comes to constructTransaction endpoint this is part of NEW transaction flow where we have construct/sign/decode/submit that contains all new features that were are in the last two years. `joinStakePool` like `postTransaction` are legacy endpoints (they have construct/sign/submit in one step). Of course I understand that we might ALSO support joinStakePool having conway voting feature as daedalus might use it (although not sure if they have not adopted new tx workflow). Nevertheless, I do think voting should be primarly supported in constructTransaction.... Although I understand that using `handleVotingWhenMissingInConway` might be too restrictive for the endpoint.

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 9 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: When it comes to constructTransaction endpoint this is part of NEW transaction flow where we have construct/sign/decode/submit that contains all new features that were are in the last two years. joinStakePool like postTransaction are legacy endpoints (they have construct/sign/submit in one step). Of course I understand that we might ALSO support joinStakePool having conway voting feature as daedalus might use it (although not sure if they have not adopted new tx workflow). Nevertheless, I do think voting should be primarly supported in constructTransaction....

File: specifications/Cardano/Wallet/delegation.md
Comment (paweljakubas) [reply specifications/Cardano/Wallet/delegation.md:1]: sure, done here https://github.com/cardano-foundation/cardano-wallet/pull/4444
commit removed in this PR

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 9 above)
Comment (HeinrichApfelmus) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: > Nevertheless, I do think voting should be primarly supported in constructTransaction....

Yes, I do agree with that — what I'm saying is that in `constructTransaction`, voting should be explicit rather than automatic/implicit. In other words, `constructTransaction` should not implicitly vote `Abstain` if the JSON data does not explicitly contain a vote to `abstain`.

> Of course I understand that we might ALSO support joinStakePool having conway voting feature as daedalus might use it (although not sure if they have not adopted new tx workflow).

Yes. I believe that Daedalus is still using `joinStakePool`, but worth checking.

File: lib/wallet/src/Cardano/Wallet/Transaction/Delegation.hs (hunk 11 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/Transaction/Delegation.hs]: yeah, my bad :smiling_face_with_tear:  I have added `Delegation.Voting` module

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs (hunk 12 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs]: done. I do think `nub $ delegCerts <> votingCerts` should be enough as we use the same staking key during registration

Comment (HeinrichApfelmus) [reply lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs]: > I do think nub $ delegCerts <> votingCerts should be enough as we use the same staking key during registration

Ah, `nub` is a good idea.

One small hitch: We have to careful whether the certificates are applied left-to-right, or right-to-left. In the situation

```hs
  nub ["Delegate", "Register", "Vote", "Register"]
=
  nub ["Delegate", "Register", "Vote"] 
```

applying right-to-left would be a mistake, because `nub` takes the leftmost element when deduplicating.

**Comment by paweljakubas** [comment]
@HeinrichApfelmus joinStakePool and quitStakePool logic of vote to abstain added and removed from constructTransaction -> https://github.com/cardano-foundation/cardano-wallet/pull/4438/commits/e2627af6f671c7c69c9d87cf19d64f1288d3108f

File: lib/wallet/src/Cardano/Wallet.hs [hunk 13]
```diff
@@ -2646,6 +2657,65 @@ submitExternalTx tr nw tl sealedTx = do
...
+    -> DBLayer IO s
+    -> Bool
+    -> IO (Maybe VotingAction)
+handleVotingWhenMissingInConway era db delegationsPresent = do
+    areWeInConway <- case votingEnabledInEra era of
+        Left _ -> pure False
+        Right _ -> pure True
+    voting <- haveWeVoted db
+    stakingKeyRegistered <- isStakeKeyInDb db
+    if (areWeInConway && not voting && delegationsPresent) then
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet.hs]:
```suggestion
    -> IO (Maybe VotingAction)
handleVotingWhenMissingInConway era db = do
    areWeInConway <- case votingEnabledInEra era of
        Left _ -> pure False
        Right _ -> pure True
    voting <- haveWeVoted db
    stakingKeyRegistered <- isStakeKeyInDb db
    if (areWeInConway && not voting) then
```

After removing the `handleVotingWhenMissingInConway` from `constructTransaction` and making it part of `selectCoinsForJoin`, the `delegationsPresent` parameter is now always `True` — and we can remove it. 🤓

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 14]
```diff
@@ -2210,9 +2219,14 @@ selectCoinsForQuit ctx@ApiLayer{..} (ApiT walletId) = do
...
             db currentEpochSlotting withdrawal
         let changeAddrGen = W.defaultChangeAddressGen (delegationAddressS @n)
+
+        optionalVoteAction <-
+            liftIO $ W.handleVotingWhenMissingInConway era db True
+
         let txCtx = defaultTransactionCtx
                 { txDelegationAction = Just action
                 , txWithdrawal = withdrawal
+                , txVotingAction = optionalVoteAction
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion

        let txCtx = defaultTransactionCtx
                { txDelegationAction = Just action
                , txWithdrawal = withdrawal
```

We do not need to supply a `txVotingAction ≠ Nothing` here, because the result of `WD.quitStakePoolDelegationAction` can only be `Quit`. There is no point in adding a vote to `Abstain` if we want to quit anyway. 🤔

We do need the `txDeposit`, however — the code that constructs the certificate corresponding to the `Quit` action will refuse to do so without a deposit size. 😲

File: lib/wallet/src/Cardano/Wallet/Transaction/Delegation.hs [hunk 15]
```diff
@@ -56,11 +63,13 @@ certificateFromDelegationAction
         -- ^ Era in which we create the certificate
     -> Either XPub (Script KeyHash)
         -- ^ Our staking credential
+    -> Maybe Coin
+        -- ^ Optional deposit value
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Transaction/Delegation.hs:67]:
```
    -> Coin
        -- ^ Deposit value
```

Ideally, we want this to be `Coin` rather than `Maybe Coin` — the deposit is always known from the protocol parameters. However, fixing this in the current codebase is hard — the best approach would be to carry around the protocol parameters around all the way down `mkUnsignedTransaction` or even further.

File: lib/wallet/src/Cardano/Wallet/Transaction/Voting.hs [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+    ( toShort
+    )
+
+import qualified Cardano.Api as Cardano
+import qualified Cardano.Api.ReexposeLedger as Ledger
+import qualified Cardano.Api.Shelley as Cardano
+import qualified Internal.Cardano.Write.Tx as Write
+    ( CardanoApiEra
+    , RecentEra (RecentEraBabbage, RecentEraConway)
+    )
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Transaction/Voting.hs]:
```suggestion
```

(🤷 I see no value in adding explicit exports to qualified imports.)

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 17]
```diff
@@ -728,22 +749,25 @@ mkUnsignedTransaction networkId stakeCred ctx selection = do
...
-                        certificateFromDelegationAction era (Right script) action
+                        certificateFromDelegationAction era (Right script)
+                        depositM action
                     Right Nothing ->
                         error $ unwords
                             [ "stakeCred in mkUnsignedTransaction must be"
                             , "either xpub or script when there is delegation"
                             , "action"
                             ]
+            let certs = L.nub $ votingCerts <> delegCerts
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs:770]: 👍  I have checked. In Conway, the rule handling the certificates is called `CERTS`. As far as I can tell, it goes left to right, so `nub` is appropriate here.

**Review by HeinrichApfelmus — APPROVED** [review]

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 14 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: indeed, txDeposit is needed when registration and derigistration happens.  no optionalVoteAction during quitting stake pool!

Comment (HeinrichApfelmus) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: > I do think when quitPools happens and there is voting then we need to emit user error ....

As the "Quit" endpoint goes, I think it's better to not emit an error when the wallet is voting, because this is a legacy endpoint and doesn't expect anything about voting anyway.

As far as this function goes, I think this would not happen anyway? 🤔 

* `handleVotingWhenMissingInConway` doesn't emit an error.
* The only reason why there may be an error is when `txDeposit = Nothing` and `txDelegationAction = Just Quit` and the function `certificateFromDelegationAction ` will hit the following case:

```hs
       (Quit, Nothing) ->
           error "certificateFromDelegationAction: deposit value required in \
                 \Conway era when registration is carried out (quitting)"
```

I would like to not hit that case, because that's a bug — we didn't provide the deposit value.

Comment (Anviking) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: > txDeposit is needed when registration and derigistration happens

Related: https://cardanofoundation.atlassian.net/browse/ADP-3270

To properly handle:

```haskell
-- | De-Register the staking credential. Deposit, if present, must match the amount
-- that was left as a deposit upon stake credential registration.
```

https://github.com/IntersectMBO/cardano-ledger/blob/53a64c898408966940c66fa4b29d42f2656240f4/eras/conway/impl/src/Cardano/Ledger/Conway/TxCert.hs#L391-L392

i.e., the deposit not necessarily being `currentPP ^. ppKeyDepositL`, I believe we'd need to either get it over `LSQ` if possible or track it in our DB.

File: lib/wallet/src/Cardano/Wallet/Transaction/Delegation.hs (hunk 15 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/Transaction/Delegation.hs:67]: I will leave it for separate PR

File: lib/wallet/src/Cardano/Wallet/Transaction/Voting.hs (hunk 16 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/Transaction/Voting.hs]: yep, the same was in Transaction.Delegation module and I automatically copy pasted it. Now removed that strange way of importing in both modules

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 14 above)
Comment (HeinrichApfelmus) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: > track it in our DB.

😵‍💫 I think that's what we have to do — but perhaps not as part of the Conway hardfork.

The ledger still recognizes the old way of unregistering certificates, it's just that `Cardano.Api` does not expose it as part of `makeStakeAddressUnregistrationCertificate`. It's possible to use the constructor `ConwayCertificate` directly, though.
